### ISSUE TITLE
Crash when switching git branches

### ISSUE THREAD

**Issue by Noxigar** [MBulli/SmartCommandlineArgs#43]
The version 2.0.2 is crashing frequently when I'm changing branches or when I'm merging branches.

Microsoft Visual Studio Enterprise 2017 
Version 15.5.2
VisualStudio.15.Release/15.5.2+27130.2010
Microsoft .NET Framework
Version 4.7.02556

Installed Version: Enterprise

Architecture Diagrams and Analysis Tools   00369-60000-00001-AA455
Microsoft Architecture Diagrams and Analysis Tools

Visual Basic 2017   00369-60000-00001-AA455
Microsoft Visual Basic 2017

Visual C# 2017   00369-60000-00001-AA455
Microsoft Visual C# 2017

Visual C++ 2017   00369-60000-00001-AA455
Microsoft Visual C++ 2017

ASP.NET and Web Tools 2017   15.0.31125.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   1.0
Provides languages services for ASP.NET Core Razor.

Child Process Debugging Power Tool   1.0
Power tool to add child process debugging to Visual Studio.

ClaudiaIDE   1.28.5
This extension change the background image of editor.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

CustomizeVSWindowTitle 3   1.0
Information about my package

File Icons   2.7
Adds icons for files that are not recognized by Solution Explorer

GitFlow.VS.Extension   1.0
Visual Studio extension that integrates GitFlow

Image Watch   1.0

JavaScript Language Service   2.0
JavaScript Language Service

Markdown Editor   1.11.214
A full featured Markdown editor with live preview and syntax highlighting. Supports GitHub flavored Markdown.

MedeosDebugVisualizer2013   1.0
Information about my package

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   4.5.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

Open Command Line   2.1.179
Opens a command line at the root of the project. Support for all consoles such as CMD, PowerShell, Bash etc. Provides syntax highlighting, Intellisense and execution of .cmd and .bat files.

Open in Notepad++   1.1.12
Adds a menu command that lets you open any solution, project, folder and file in Notepad++

Smart Commandline Arguments Extension   1.5.1
A Visual Studio Extension which aims to provide a better UI to manage your command line arguments. For additional information, visit https://github.com/MBulli/SmartCommandlineArgs

StopOnFirstBuildError   1.0
StopOnFirstBuildError

Stops a solution build immediately after a project has failed to build.

http://tech.einaregilsson.com/2011/01/06/stop-build-on-first-error-in-visual-studio-2010/

Switch Startup Project   3.5.39
Provides a toolbar dropdown box  to switch between startup projects.

Syntax Highlighting Pack   2.7.125
Adds syntax highlighting and snippet support for a wide variety of programming languages such as Clojure, Go, Jade, Lua, Swift, Ruby and many more...

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

Visual Assist   
For more information about Visual Assist, see the Whole Tomato Software website at http://www.WholeTomato.com. Copyright (c) 1997-2017 Whole Tomato Software, Inc.

Visual C++ for Linux Development   1.0.8
Visual C++ for Linux Development

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

VisualGDB   5.3
Allows developing and debugging Embedded, Linux, Android and other GCC/GDB-based applications with Visual Studio.

VSColorOutput   2.5.1
Color output for build and debug windows - http://mike-ward.net/vscoloroutput

WorkingFilesWindow Extension   1.0
WorkingFilesWindow Visual Studio Extension Detailed Info

Workspace Manager Extension   1.0
Workspace Manager for Visual Studio allows to save, load and merge user workspaces for each solution.

**Comment by MBulli** [comment]
Well, I found a bug which was caused in conjunction with git checkout.
A race condition occurred if a json file has different content across the branches and the solution is reloaded after the checkout.
I'm not sure if that's your problem. That's why I'm posting a test version here.
Please give it a try and let me know if the problem is gone.

[SmartCmdArgs-v2.0.2.1.zip](https://github.com/MBulli/SmartCommandlineArgs/files/1636928/SmartCmdArgs-v2.0.2.1.zip)

**Comment by Noxigar** [comment]
Hello, 

It's seems to be my exact case, I'll will try this build tomorrow, thanks a lot.

**Comment by Noxigar** [comment]
Good news, the crash is gone.
One other problem, it's when i change my branch the file solutionname.args.json is overridden by the file of the previous branch.
Perhaps it's by design, but as we have the choice to store the arguments into a git repository, I'm feeling that the arguments be reloaded after a git checkout (and solution reload), not to override the new checkout with the another branch.   

If I perform 'undo checkout' on the solutionname.args.json, the good content is displayed.

**Comment by MBulli** [comment]
I also noticed this when I tried to reproduce the crash.
However, today I was unable to trigger that behavior.
Does it happen all the time?

The following simple test just worked as expected:
```
MINGW64 /d/Temp/tests/ConsoleAppCsharpGit/ConsoleAppCsharpGit (master)
$ git status
On branch master
nothing to commit, working tree clean

MINGW64 /d/Temp/tests/ConsoleAppCsharpGit/ConsoleAppCsharpGit (master)
$ cat ConsoleAppCsharpGit.args.json
{
  "FileVersion": 2,
  "Id": "82826d4b-06ad-4fe9-b136-291adf0ff56d",
  "Items": [
    {
      "Id": "9aa8b591-2d87-4c6d-a169-43d494a58ba2",
      "Command": "branch master"
    }
  ]
}
MINGW64 /d/Temp/tests/ConsoleAppCsharpGit/ConsoleAppCsharpGit (master)
$ git checkout test
Switched to branch 'test'

MINGW64 /d/Temp/tests/ConsoleAppCsharpGit/ConsoleAppCsharpGit (test)
$ git status
On branch test
nothing to commit, working tree clean

MINGW64 /d/Temp/tests/ConsoleAppCsharpGit/ConsoleAppCsharpGit (test)
$ cat ConsoleAppCsharpGit.args.json
{
  "FileVersion": 2,
  "Id": "82826d4b-06ad-4fe9-b136-291adf0ff56d",
  "Items": [
    {
      "Id": "9aa8b591-2d87-4c6d-a169-43d494a58ba2",
      "Command": "branch test"
    }
  ]
}
MINGW64 /d/Temp/tests/ConsoleAppCsharpGit/ConsoleAppCsharpGit (test)
$
```

**Comment by Noxigar** [comment]
It"s happen all the time, but perhaps it's because all my args.json are not in FileVersion "2" in the repository.
I'll check that.

**Comment by Noxigar** [comment]
After some testing with two args.json in FileVersion "2", the problem remains.
When I change my branch VS2017 is always performing a solution load, it's seems that in this case the arguments are not unloaded or reloaded from the new args.json file.

**Comment by MBulli** [comment]
Okay, this suggest that the solution reload causes the problems. I'm going to check that tonight.

**Comment by MBulli** [comment]
I spotted the problem. VS calls git to change the branch. At this time the json file has the 'right' content. However, right after git is done VS calls OnSaveOptions() where we save our arguments to disk. That's why the file gets overwritten.
For now I don't have an idea to fix that serious problem,

**Comment by Noxigar** [comment]
I suppose there no way to load the data from the disk just after the git branch change and before the SaveOptions() call ? I'll check in the VSSDK if i can find such a called method. Do you have another source of information apart the VSSDK ?

**Comment by MBulli** [comment]
Another fix could be to immediately save changes to file as soon as they occur. This would allow us to remove the harmful `SaveJsonForProject()` in `OnSaveOptions()`.

**Comment by Noxigar** [comment]
Yes, it's seems a good idea, but it need some testing I suppose.

**Comment by MBulli** [comment]
Just to keep you updated. We were quite busy but today I found some time to get back to the extension. It took me a while to gather all the events from our tree but eventually I found a nice way. Now I need to update the json and soon you can give it a try.

**Comment by MBulli** [comment]
Here's the first try to save changes immediately to disk.
Right know the move item up and down buttons are not causing a save operation.

It should work but there is some fine tuning required.
Unfortunately I hardly have a free mind to spend time working on the extension but I hope this will change soon.

[SmartCmdArgs-v2.0.2.2.zip](https://github.com/MBulli/SmartCommandlineArgs/files/1700755/SmartCmdArgs-v2.0.2.2.zip)

**Comment by Noxigar** [comment]
Thanks a lot for your time on this extension. I'll test this version tomorrow.

**Comment by Noxigar** [comment]
This build is fine for me, the .json file is now correctly handled when switching branches.

**Comment by MBulli** [comment]
@Noxigar FYI: I just pushed a new version to VS Marketplace. This version also includes the fix for your issue. I'll leave it open here because we still need to fix the move up/down buttons
